### DISCUSSION TITLE
Setting callbacks for server programs.

### DISCUSSION THREAD

**Discussion by kohnish** [joaotavora/eglot#1302] [Q&A] [answered]
Hi, I'm working in a crappy project where multiple languages are used and project root is inconsistent by design.  
  
The workaround I had in another lsp client was that to pass callback that returns the command string and let it get evaluated when a file is opened.  
  
Would that be possible to support such a thing with eglot?

**Comment by joaotavora** [comment]
From the little I was able to understand from your description it could be that you just need to tweak `project-find-functions` so that it returns the correct root in each file of the project. Then you'll be able to `M-x eglot` in each subset of files of the project that belong to a given project root.

**Reply by kohnish** [reply]
It works in the case when the PWD is the only parameter, also if I open another file in the same project separately, eglot doesn't seem to detect that the language server is already running, it works ok if I jump to file via eglot features like find-references.  
  
Language servers like eclipse-jdtls and clangd support options -data and --compile-commands-dir respectively. It'd be nice if I could modify those options per project.

**Reply by joaotavora** [reply, marked as answer]
> It'd be nice if I could modify those options per project.

Have you looked in the manual? There's a section for that. If the options can only be given as a server invocation, then directory-local eglot-server-programs is what you want.

**Reply by joaotavora** [reply]
> also if I open another file in the same project separately, eglot doesn't seem to detect that the language server is already running

This is not what should happen. Eglot should most definitely reuse a language server for files of the same major mode in the same project. This is described in the manual too.

**Reply by kohnish** [reply]
> directory-local eglot-server-programs  
  

Oh. Sorry, I must've missed it. Let me try that instead.
Thank you so much for prompt response.
